XIPs/xip-31-message-backup-request-system.md

```diff
@@ -0,0 +1,466 @@
+
+   6d. Load each message into the local database, ignoring any duplicate messages
+
+   6e. Delete the `MessageHistoryBackupResponse` and all associated credentials
+
+   6f. Set the status of the `MessageHistoryBackupRequest` to `Applied`
+
+#### Flow for Message Backup Providers
+
+1. Receive a `MessageHistoryBackupRequest` as part of normal message receiving flow
+1. Retrieve the contact for the installation that sent the message, and ensure that it has not been revoked. If revoked, ignore.
+1. Ensure that the installation that sent the message has a contact signed by the same blockchain account as the current user. If not, ignore.
+1. Convert all messages in the local database into a Message Backup File (maybe we want to chunk here?)
+1. Generate ephemeral encryption key, salt, and nonce. Encrypt the file using these keys
+1. Upload the file to the `backupStorageProviderUploadUrl` from the request
+1. Reply to the message with a `MessageHistoryBackupResponse` containing the encryption details, the hash of the backup file, and the `backupUrl` provided by the Backup Storage Provider
+1. Delete the `MessageHistoryBackupRequest`, the local database dump, and all the encryption keys.
+
+#### End-to-end flow
+
+```mermaid
+sequenceDiagram
+    Participant B as Backup Requester
+    Participant S as Backup Storage Provider
+    Participant N as XMTP Network
+    Participant M as Message Backup Provider
+    B->>S: Request a backup storage upload URL
+    S-->>B: Receive one time upload URL
+    B->>N: Send a MessageHistoryBackupRequest to each potential Message Backup Provider
+    M->>N: Check for MessageHistoryBackupRequest
+    N-->>M: Receive MessageHistoryBackupRequest
+    M->>S: Upload backup file
+    M->>N: Send MessageHistoryBackupResponse
+    B->>N: Check for MessageHistoryBackupResponse
+    N-->>B: Receive response
+    B->>S: Request file from URL in response
+    S-->>B: Receive file
+    B-->>B: Decrypt and load file into database
+```
+
+### Backup Account Files
+
+Backup account files are designed as a solution for users who have lost access to all of their XMTP applications. To create a Backup Account File, an application will generate a new set of XMTP keys and register the installation on the XMTP network. The installation contact published to the network will be set with a type of `backup`, which instructs all well-behaved clients to disregard any messages from this installation other than `MessageHistoryBackupResponse`s. This means that a compromised backup installation cannot be used to send messages on the user's behalf, which makes these files a less valuable target for attack.
+
+Once a backup account's contact has been published to the network, their installation messaging topic will receive a copy of all new messages sent to the associated blockchain account.
```

Comment:
One thing to mention is that the backup client doesn't have access to historical messages. We would have to introduce new payload formats to retransmit all existing history to the backup client over the XMTP network

Reply:
That's true. It will only have a complete history of your XMTP messaging if it is created as part of your first interaction with the XMTP network. If you create it later, it will only have history starting at the time of it's creation.